src/Workspaces/Remote/Core/Diagnostics/PerformanceQueue.cs

```diff
@@ -0,0 +1,226 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote.Diagnostics
+{
+    /// <summary>
+    /// This queue hold onto raw performance data. this type itself is not thread safe. the one who uses this type
+    /// should take care of that.
+    /// </summary>
+    internal class PerformanceQueue
+    {
+        // we need at least 100 samples for result to be stable
+        private const int MinSampleSize = 100;
+
+        private readonly int _maxSampleSize;
+        private readonly LinkedList<Snapshot> _snapshots;
+
+        public PerformanceQueue(int maxSampleSize)
+        {
+            Contract.ThrowIfFalse(maxSampleSize > MinSampleSize);
+
+            _maxSampleSize = maxSampleSize;
+            _snapshots = new LinkedList<Snapshot>();
+        }
+
+        public int Count => _snapshots.Count;
+
+        public void Add(IEnumerable<(string, TimeSpan)> rawData, int unitCount)
```

Comment:
:bulb: Named tuple elements (only commenting here but there are a bunch of existing cases in this pull request where names would be helpful)